File: frontend/services/complaint.service.ts
import { Complaint, CreateComplaintData, ComplaintCategory, ComplaintUrgency, ComplaintLocation, ComplaintMedia, Comment } from "@/types";
import { apiClient } from "./api.client";
import { useAuthStore } from "@/store/useAuthStore";

// Cloudinary base URL for media files - used for prepending to relative paths
// This should be configured in your .env.local
// Example: NEXT_PUBLIC_CLOUDINARY_BASE_URL=https://res.cloudinary.com/your-cloud-name/image/upload
const getCloudinaryBaseUrl = (): string => {
  // Try environment variable first
  const envUrl = process.env.NEXT_PUBLIC_CLOUDINARY_BASE_URL;
  if (envUrl) return envUrl;
  
  // Default - will be replaced with actual uploaded URL
  return "";
};

/**
 * Helper function to get full Cloudinary URL
 * If the URL is already a full URL (starts with http), return as-is
 * Otherwise, check if it's a Cloudinary public ID and construct the URL
 */
export function getFullMediaUrl(url: string | undefined): string {
  if (!url) return "";
  
  // If already a full URL, return as-is
  if (url.startsWith("http://") || url.startsWith("https://")) {
    return url;
  }
  
  // If it's a relative path (starts with /), return as-is
  // The backend should return full URLs after upload
  if (url.startsWith("/")) {
    // Could prepend API base URL if needed
    const apiUrl = process.env.NEXT_PUBLIC_API_URL || "http://localhost:5000";
    return `${apiUrl}${url}`;
  }
  
  // If it's a Cloudinary public ID (no slashes), construct URL
  // This assumes the backend stores just the public ID
  const cloudName = process.env.NEXT_PUBLIC_CLOUDINARY_CLOUD_NAME || "demo";
  return `https://res.cloudinary.com/${cloudName}/image/upload/${url}`;
}

/**
 * Process complaint data to fix media URLs
 * This ensures all media URLs are full URLs
 */
export function processComplaintMedia(complaint: Complaint): Complaint {
  if (!complaint) return complaint;
  
  // Process media array
  if (complaint.media) {
    complaint.media = complaint.media.map(item => ({
      ...item,
      url: getFullMediaUrl(item.url)
    }));
  }
  
  // Process before photos
  if (complaint.beforePhotos) {
    complaint.beforePhotos = complaint.beforePhotos.map(photo => ({
      ...photo,
      url: getFullMediaUrl(photo.url)
    }));
  }
  
  // Process after photos
  if (complaint.afterPhotos) {
    complaint.afterPhotos = complaint.afterPhotos.map(photo => ({
      ...photo,
      url: getFullMediaUrl(photo.url)
    }));
  }
  
  return complaint;
}

/**
  * Upload media files to Cloudinary
  */
export const uploadMedia = async (
  files: File[]
): Promise<{ success: boolean; data?: ComplaintMedia[]; message?: string }> => {
  const formData = new FormData();
  files.forEach((file) => {
    formData.append('media', file);
  });

  try {
    const apiUrl = process.env.NEXT_PUBLIC_API_URL || "http://localhost:5000/api";
    
    const response = await fetch(`${apiUrl}/upload`, {
      method: 'POST',
      body: formData,
      credentials: 'include',
    });

    const result = await response.json();
    
    if (!response.ok) {
      console.error('Upload failed:', result.message);
      return { success: false, message: result.message || 'Upload failed' };
    }

    return { success: true, data: result.data };
  } catch (error) {
    console.error('Upload error:', error);
    return { success: false, message: 'Failed to upload files' };
  }
};

/**
 * Submit a new complaint (authenticated citizen)
 */
export const submitComplaint = async (
  data: CreateComplaintData
): Promise<{ message: string; complaint: Complaint }> => {
  // apiClient gère déjà le token (useAuthStore) + refresh
  return apiClient.post<{ message: string; complaint: Complaint }>("/citizen/complaints", data);
};

/**
 * Get citizen's complaints (with optional filters)
 */
export const getMyComplaints = async (params?: {
  status?: string;
  category?: string;
  page?: number;
  limit?: number;
}): Promise<{
  message: string;
  complaints: Complaint[];
  pagination: {
    total: number;
    page: number;
    limit: number;
    pages: number;
  };
}> => {
  const searchParams = new URLSearchParams();
  if (params?.status) searchParams.set("status", params.status);
  if (params?.category) searchParams.set("category", params.category);
  if (params?.page) searchParams.set("page", params.page.toString());
  if (params?.limit) searchParams.set("limit", params.limit.toString());

  const queryString = searchParams.toString();
  const endpoint = `/citizen/complaints${queryString ? `?${queryString}` : ""}`;

  console.log("Fetching from endpoint:", endpoint);
  const result = await apiClient.get<{
    message: string;
    complaints: Complaint[];
    pagination: {
      total: number;
      page: number;
      limit: number;
      pages: number;
    };
  }>(endpoint);
  console.log("API response:", result);
  return result;
};

/**
 * Get a single complaint by ID (citizen view)
 */
export const getComplaintById = async (
  id: string
): Promise<{ message: string; complaint: Complaint }> => {
  return apiClient.get<{ message: string; complaint: Complaint }>(`/citizen/complaints/${id}`);
};

/**
 * Update citizen's own complaint (only if SUBMITTED status)
 */
export const updateComplaint = async (
  id: string,
  data: Partial<{
    title: string;
    description: string;
    category: ComplaintCategory;
    urgency: ComplaintUrgency;
    location: ComplaintLocation;
    media: ComplaintMedia[];
  }>
): Promise<{ message: string; complaint: Complaint }> => {
  return apiClient.put<{ message: string; complaint: Complaint }>(`/citizen/complaints/${id}`, data);
};

/**
 * Delete citizen's own complaint (only if SUBMITTED status)
 */
export const deleteComplaint = async (
  id: string
): Promise<{ message: string }> => {
  return apiClient.delete<{ message: string }>(`/citizen/complaints/${id}`);
};

/**
 * Get complaint detail for agent/manager (BL-16)
 * Returns full complaint details including media, location, and reporter info
 */
export const getComplaintDetail = async (
  id: string
): Promise<{ success: boolean; data: Complaint }> => {
  return apiClient.get<{ success: boolean; data: Complaint }>(`/complaints/${id}`);
};

/**
 * Get all complaints (admin/agent view)
 */
export const getAllComplaints = async (params?: {
  status?: string;
  category?: string;
  governorate?: string;
  municipality?: string;
  search?: string;
  page?: number;
  limit?: number;
  includeArchived?: boolean;
}): Promise<{
  success: boolean;
  data: {
    complaints: Complaint[];
    pagination: {
      total: number;
      page: number;
      limit: number;
      pages: number;
    };
  };
}> => {
  const searchParams = new URLSearchParams();
  if (params?.status) searchParams.set("status", params.status);
  if (params?.category) searchParams.set("category", params.category);
  if (params?.governorate) searchParams.set("governorate", params.governorate);
  if (params?.municipality) searchParams.set("municipality", params.municipality);
  if (params?.search) searchParams.set("search", params.search);
  if (params?.page) searchParams.set("page", params.page.toString());
  if (params?.limit) searchParams.set("limit", params.limit.toString());
  if (params?.includeArchived) searchParams.set("includeArchived", "true");

  const queryString = searchParams.toString();
  const endpoint = `/complaints${queryString ? `?${queryString}` : ""}`;

  return apiClient.get<{
    success: boolean;
    data: {
      complaints: Complaint[];
      pagination: {
        total: number;
        page: number;
        limit: number;
        pages: number;
      };
    };
  }>(endpoint);
};

/**
 * Update complaint status (validate, reject, etc.)
 */
export const updateComplaintStatus = async (
  id: string,
  status: string,
  rejectionReason?: string
): Promise<{ success: boolean; data: Complaint; message?: string }> => {
  return apiClient.patch<{ success: boolean; data: Complaint; message?: string }>(
    `/complaints/${id}/status`,
    { status, rejectionReason }
  );
};

/**
 * Assign complaint to a technician
 */
export const assignComplaint = async (
  id: string,
  assignedToId: string
): Promise<{ success: boolean; data: Complaint; message?: string }> => {
  return apiClient.patch<{ success: boolean; data: Complaint; message?: string }>(
    `/complaints/${id}/assign`,
    { assignedToId }
  );
};

/**
 * Assign complaint to a department
 */
export const updateComplaintDepartment = async (
  id: string,
  departmentId: string
): Promise<{ success: boolean; data: Complaint; message?: string }> => {
  return apiClient.patch<{ success: boolean; data: Complaint; message?: string }>(
    `/complaints/${id}/department`,
    { departmentId }
  );
};

/**
 * Update complaint priority/urgency
 */
export const updateComplaintPriority = async (
  id: string,
  urgency: string,
  priorityScore?: number
): Promise<{ success: boolean; data: Complaint; message?: string }> => {
  return apiClient.patch<{ success: boolean; data: Complaint; message?: string }>(
    `/complaints/${id}/priority`,
    { urgency, priorityScore }
  );
};

/**
 * Add a comment/note to a complaint
 * isInternal: true for internal notes (staff only)
 */
export const addComplaintComment = async (
  id: string,
  text: string,
  isInternal: boolean = false
): Promise<{ success: boolean; data: Comment; message?: string }> => {
  return apiClient.post<{ success: boolean; data: Comment; message?: string }>(
    `/complaints/${id}/comments`,
    { text, isInternal }
  );
};

/**
 * Citizen confirms resolution (changes RESOLVED to CLOSED)
 */
export const confirmResolution = async (
  id: string
): Promise<{ success: boolean; data: Complaint; message?: string }> => {
  return apiClient.patch<{ success: boolean; data: Complaint; message?: string }>(
    `/complaints/${id}/status`,
    { status: "CLOSED" }
  );
};

/**
 * Archive a complaint (admin only)
 */
export const archiveComplaint = async (
  id: string
): Promise<{ success: boolean; data: Complaint; message?: string }> => {
  return apiClient.patch<{ success: boolean; data: Complaint; message?: string }>(
    `/complaints/${id}/archive`,
    {}
  );
};

/**
 * Unarchive a complaint (admin only)
 */
export const unarchiveComplaint = async (
  id: string
): Promise<{ success: boolean; data: Complaint; message?: string }> => {
  return apiClient.patch<{ success: boolean; data: Complaint; message?: string }>(
    `/complaints/${id}/unarchive`,
    {}
  );
};

/**
 * Get archived complaints (admin view)
 */
export const getArchivedComplaints = async (params?: {
  filter?: string;
  search?: string;
  page?: number;
  limit?: number;
}): Promise<{
  success: boolean;
  complaints: Complaint[];
  total: number;
  page: number;
  pages: number;
}> => {
  const searchParams = new URLSearchParams();
  if (params?.filter) searchParams.set('filter', params.filter);
  if (params?.search) searchParams.set('search', params.search);
  if (params?.page) searchParams.set('page', params.page.toString());
  if (params?.limit) searchParams.set('limit', params.limit.toString());

  const queryString = searchParams.toString();
  const endpoint = `/complaints/archived${queryString ? `?${queryString}` : ''}`;

  return apiClient.get<{
    success: boolean;
    complaints: Complaint[];
    total: number;
    page: number;
    pages: number;
  }>(endpoint);
};

/**
 * Predict category using AI
 */
export const predictCategory = async (text: string): Promise<{
  predicted: string;
  confidence: number;
  alternatives: string[];
  reasoning: string;
}> => {
  const apiUrl = process.env.NEXT_PUBLIC_API_URL || "http://localhost:5000";
  
  const response = await fetch(`${apiUrl}/api/ai/predict-category`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    body: JSON.stringify({ text }),
  });

  if (!response.ok) {
    return {
      predicted: "AUTRE",
      confidence: 0,
      alternatives: [],
      reasoning: "AI service unavailable",
    };
  }

  return response.json();
};

/**
 * Extract keywords using AI
 */
export const extractKeywords = async (text: string): Promise<{
  keywords: string[];
  locationKeywords: string[];
  urgencyKeywords: string[];
  similarityHash: string;
}> => {
  const aiUrl = process.env.NEXT_PUBLIC_AI_SERVICE_URL || "http://localhost:8000";
  
  const response = await fetch(`${aiUrl}/ai/extract-keywords`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ text }),
  });

  if (!response.ok) {
    return {
      keywords: [],
      locationKeywords: [],
      urgencyKeywords: [],
      similarityHash: "",
    };
  }

  return response.json();
};

/**
 * Confirm a complaint (citizen BL-28)
 * Returns updated confirmation count
 */
export const confirmComplaint = async (
  id: string
): Promise<{ success: boolean; confirmationCount: number; message: string }> => {
  return apiClient.post<{ success: boolean; confirmationCount: number; message: string }>(
    `/complaints/${id}/confirm`,
    {}
  );
};

/**
 * Remove confirmation from a complaint (citizen BL-28)
 * Only allowed within 24 hours
 */
export const unconfirmComplaint = async (
  id: string
): Promise<{ success: boolean; confirmationCount: number; message: string }> => {
  return apiClient.delete<{ success: boolean; confirmationCount: number; message: string }>(
    `/complaints/${id}/confirm`
  );
};

/**
 * Upvote a complaint (citizen BL-28)
 * Returns updated upvote count
 */
export const upvoteComplaint = async (
  id: string
): Promise<{ success: boolean; upvoteCount: number; message: string }> => {
  return apiClient.post<{ success: boolean; upvoteCount: number; message: string }>(
    `/complaints/${id}/upvote`,
    {}
  );
};

/**
 * Remove upvote from a complaint (citizen BL-28)
 * Only allowed within 24 hours
 */
export const removeUpvote = async (
  id: string
): Promise<{ success: boolean; upvoteCount: number; message: string }> => {
  return apiClient.delete<{ success: boolean; upvoteCount: number; message: string }>(
    `/complaints/${id}/upvote`
  );
};

export const complaintService = {
  submitComplaint,
  getMyComplaints,
  getComplaintById,
  updateComplaint,
  deleteComplaint,
  getComplaintDetail,
  getAllComplaints,
  updateComplaintStatus,
  assignComplaint,
  updateComplaintDepartment,
  updateComplaintPriority,
  addComplaintComment,
  confirmResolution,
  archiveComplaint,
  unarchiveComplaint,
  getArchivedComplaints,
  getFullMediaUrl,
  processComplaintMedia,
  predictCategory,
  extractKeywords,
  confirmComplaint,
  unconfirmComplaint,
  upvoteComplaint,
  removeUpvote,
};

File: mobile/lib/services/complaint_service.dart
import 'api_client.dart';
import '../models/complaint_model.dart';

class ComplaintService {
  final ApiClient _apiClient = ApiClient();

  // ─── Citizen ───

  Future<List<Complaint>> getMyComplaints({
    int page = 1,
    int limit = 20,
  }) async {
    final response = await _apiClient.get(
      '/citizen/complaints?page=$page&limit=$limit',
    );
    List<dynamic> list = [];
    if (response is Map) {
      final data = response['data'];
      if (data is Map) {
        list = (data['complaints'] as List?) ?? [];
      } else {
        list = (response['complaints'] as List?) ?? [];
      }
    }
    return list.map((c) => Complaint.fromJson(c as Map<String, dynamic>)).toList();
  }

  Future<Complaint> getComplaintById(String id) async {
    try {
      final response = await _apiClient.get('/citizen/complaints/$id');
      return Complaint.fromJson(response['complaint'] ?? response['data'] ?? {});
    } catch (_) {
      final publicResponse = await _apiClient.get('/public/complaints/$id');
      return Complaint.fromJson(
        publicResponse['complaint'] ?? publicResponse['data'] ?? {},
      );
    }
  }

  Future<Map<String, dynamic>> createComplaint(
    Map<String, dynamic> data,
  ) async {
    final response = await _apiClient.post('/citizen/complaints', data);
    return response['complaint'] ?? response['data'] ?? {};
  }

  Future<Map<String, dynamic>> updateComplaint(
    String id,
    Map<String, dynamic> data,
  ) async {
    final response = await _apiClient.put('/citizen/complaints/$id', data);
    return response['complaint'] ?? {};
  }

  Future<void> deleteComplaint(String id) async {
    await _apiClient.delete('/citizen/complaints/$id');
  }

  Future<Map<String, dynamic>> getCitizenStats() async {
    final response = await _apiClient.get('/citizen/stats');
    return _extractData(response);
  }

  Future<Map<String, dynamic>> getAgentStats() async {
    final response = await _apiClient.get('/agent/stats');
    return _extractData(response);
  }

  Future<Map<String, dynamic>> getManagerStats() async {
    final response = await _apiClient.get('/manager/stats');
    return _extractData(response);
  }

  Future<Map<String, dynamic>> getTechnicianStats() async {
    final response = await _apiClient.get('/technician/stats');
    return _extractData(response);
  }

  Future<Map<String, dynamic>> getAdminStats() async {
    final response = await _apiClient.get('/admin/stats');
    return _extractData(response);
  }

  /// Safely extract data from an API response.
  Map<String, dynamic> _extractData(dynamic response) {
    if (response is Map) {
      final data = response['data'];
      if (data is Map<String, dynamic>) return data;
      if (data is Map) return Map<String, dynamic>.from(data);
      // Response itself might be the data
      if (response is Map<String, dynamic>) return response;
      return Map<String, dynamic>.from(response);
    }
    return {};
  }

  // ─── Confirm / Upvote ───

  Future<void> confirmComplaint(String id) async {
    await _apiClient.post('/complaints/$id/confirm', {});
  }

  Future<void> removeComplaintConfirmation(String id) async {
    await _apiClient.delete('/complaints/$id/confirm');
  }

  Future<void> upvoteComplaint(String id) async {
    await _apiClient.post('/public/complaints/$id/upvote', {});
  }

  Future<List<dynamic>> getPublicComments(String id) async {
    final response = await _apiClient.get('/public/complaints/$id/comments');
    return response['data'] ?? response['comments'] ?? [];
  }

  Future<void> addPublicComment(String id, String text, {bool anonymous = false}) async {
    await _apiClient.post('/public/complaints/$id/comment', {
      'text': text,
      'anonymous': anonymous,
    });
  }

  Future<void> submitRating(String id, int rating, String comment, bool resolvedCorrectly) async {
    await _apiClient.post('/citizen/complaints/$id/rating', {
      'rating': rating,
      'comment': comment,
      'resolvedCorrectly': resolvedCorrectly,
    });
  }

  Future<void> confirmResolution(String id) async {
    await _apiClient.post('/citizen/complaints/$id/confirm-resolution', {});
  }

  Future<Map<String, dynamic>> predictCategory(String description) async {
    final response = await _apiClient.post('/ai/predict-category', {
      'description': description,
    });
    return response['data'] ?? {};
  }

  Future<Map<String, dynamic>> checkDuplicates(Map<String, dynamic> data) async {
    final response = await _apiClient.post('/ai/check-duplicates', data);
    return response['data'] ?? {};
  }

  Future<Map<String, dynamic>> predictUrgency(Map<String, dynamic> data) async {
    final response = await _apiClient.post('/ai/predict-urgency', data);
    return response['data'] ?? {};
  }

  // ─── Agent ───

  Future<List<Complaint>> getAgentComplaints({
    String? status,
    int page = 1,
    int limit = 50,
  }) async {
    String endpoint = '/agent/complaints?page=$page&limit=$limit';
    if (status != null && status.isNotEmpty && status != 'ALL') {
      endpoint += '&status=$status';
    }
    final response = await _apiClient.get(endpoint);
    final data = response['data'];
    final list = data is Map
        ? (data['complaints'] ?? [])
        : (response['complaints'] ?? []);
    return (list as List).map((c) => Complaint.fromJson(c)).toList();
  }

  Future<void> validateComplaint(String id) async {
    await _apiClient.put('/agent/complaints/$id/validate', {});
  }

  Future<void> rejectComplaint(String id, String reason) async {
    await _apiClient.put('/agent/complaints/$id/reject', {'reason': reason});
  }

  Future<void> assignDepartment(String id, String departmentId) async {
    await _apiClient.put('/agent/complaints/$id/assign-department', {
      'departmentId': departmentId,
    });
  }

  Future<void> closeComplaint(String id) async {
    await _apiClient.put('/agent/complaints/$id/close', {});
  }

  Future<void> approveResolution(String id, {String? notes}) async {
    await _apiClient.post('/agent/complaints/$id/approve-resolution', {
      if (notes != null && notes.trim().isNotEmpty) 'notes': notes,
    });
  }

  Future<void> rejectResolution(String id, {String? reason}) async {
    await _apiClient.post('/agent/complaints/$id/reject-resolution', {
      if (reason != null && reason.trim().isNotEmpty) 'reason': reason,
    });
  }

  Future<List<dynamic>> getAgentDepartments() async {
    final response = await _apiClient.get('/agent/departments');
    return response['data'] ?? response['departments'] ?? [];
  }

  // ─── Technician ───

  Future<List<Complaint>> getTechnicianTasks({
    String? status,
    int page = 1,
    int limit = 50,
  }) async {
    String endpoint = '/technician/complaints?page=$page&limit=$limit';
    if (status != null && status.isNotEmpty && status != 'ALL') {
      endpoint += '&status=$status';
    }
    final response = await _apiClient.get(endpoint);
    final data = response['data'];
    final list = data is Map
        ? (data['complaints'] ?? [])
        : (response['complaints'] ?? []);
    return (list as List).map((c) => Complaint.fromJson(c)).toList();
  }

  Future<Complaint> getTaskById(String id) async {
    final response = await _apiClient.get('/technician/complaints/$id');
    return Complaint.fromJson(response['data'] ?? {});
  }

  Future<void> startTask(String id) async {
    await _apiClient.put('/technician/complaints/$id/start', {});
  }

  Future<void> completeTask(String id, Map<String, dynamic> data) async {
    await _apiClient.put('/technician/complaints/$id/complete', data);
  }

  Future<void> addBeforePhoto(String id, List<String> urls) async {
    await _apiClient.post('/technician/complaints/$id/before-photo', {
      'photos': urls,
    });
  }

  Future<void> addAfterPhoto(String id, List<String> urls) async {
    await _apiClient.post('/technician/complaints/$id/after-photo', {
      'photos': urls,
    });
  }

  Future<void> addTechnicianComment(String id, String text) async {
    await _apiClient.post('/technician/complaints/$id/comments', {
      'text': text,
      'type': 'NOTE',
    });
  }

  // ─── Manager ───

  Future<List<Complaint>> getManagerComplaints({
    String? status,
    int page = 1,
    int limit = 50,
  }) async {
    String endpoint = '/manager/complaints?page=$page&limit=$limit';
    if (status != null && status.isNotEmpty && status != 'ALL') {
      endpoint += '&status=$status';
    }
    final response = await _apiClient.get(endpoint);
    final data = response['data'];
    final list = data is Map
        ? (data['complaints'] ?? [])
        : (response['complaints'] ?? []);
    return (list as List).map((c) => Complaint.fromJson(c)).toList();
  }

  Future<void> updatePriority(String id, String urgency) async {
    await _apiClient.put('/manager/complaints/$id/priority', {
      'urgency': urgency,
    });
  }

  Future<void> assignTechnician(String complaintId, String technicianId) async {
    await _apiClient.put('/manager/complaints/$complaintId/assign-technician', {
      'technicianId': technicianId,
    });
  }

  Future<List<dynamic>> getDepartmentTechnicians() async {
    final response = await _apiClient.get('/manager/technicians');
    return response['data'] ?? response['technicians'] ?? [];
  }

  Future<List<dynamic>> getTechnicianPerformance() async {
    final response = await _apiClient.get('/manager/technicians/performance');
    return response['data'] ?? [];
  }

  // ─── Admin ───

  Future<List<Complaint>> getAllComplaints({
    String? status,
    String? category,
    String? priority,
    String? governorate,
    String? municipality,
    String? search,
    int page = 1,
    int limit = 50,
  }) async {
    String endpoint = '/complaints?page=$page&limit=$limit';
    if (status != null && status.isNotEmpty) endpoint += '&status=$status';
    if (category != null && category.isNotEmpty) {
      endpoint += '&category=$category';
    }
    if (priority != null && priority.isNotEmpty) {
      endpoint += '&priority=$priority';
    }
    if (governorate != null && governorate.isNotEmpty) {
      endpoint += '&governorate=$governorate';
    }
    if (municipality != null && municipality.isNotEmpty) {
      endpoint += '&municipality=$municipality';
    }
    if (search != null && search.isNotEmpty) endpoint += '&search=$search';
    final response = await _apiClient.get(endpoint);
    final data = response['data'];
    final list = data is Map
        ? (data['complaints'] ?? [])
        : (response['complaints'] ?? []);
    return (list as List).map((c) => Complaint.fromJson(c)).toList();
  }

  // ─── Archive ───

  Future<List<Complaint>> getArchivedComplaints({
    int page = 1,
    int limit = 20,
  }) async {
    final response = await _apiClient.get(
      '/complaints/archived?page=$page&limit=$limit',
    );
    final list = response['complaints'] ?? [];
    return (list as List).map((c) => Complaint.fromJson(c)).toList();
  }

  // ─── Public / Transparency ───

  Future<Map<String, dynamic>> getPublicStats() async {
    final response = await _apiClient.get('/public/stats');
    if (response is Map) {
      // Try response['data'] first (standard API wrapper)
      final data = response['data'];
      if (data is Map<String, dynamic>) return data;
      if (data is Map) return Map<String, dynamic>.from(data);
      // Fallback: response itself might be the stats object
      if (response['total'] != null || response['resolved'] != null) {
        return Map<String, dynamic>.from(response);
      }
    }
    return {};
  }

  Future<List<Complaint>> getPublicComplaints({
    String? governorate,
    String? category,
    int page = 1,
    int limit = 20,
  }) async {
    String endpoint = '/public/complaints?page=$page&limit=$limit';
    if (governorate != null && governorate.isNotEmpty) {
      endpoint += '&governorate=$governorate';
    }
    if (category != null && category.isNotEmpty) {
      endpoint += '&category=$category';
    }
    final response = await _apiClient.get(endpoint);
    List<dynamic> list = [];
    if (response is Map) {
      final data = response['data'];
      if (data is Map) {
        list = (data['complaints'] as List?) ?? [];
      } else if (data is List) {
        list = data;
      } else {
        list = (response['complaints'] as List?) ?? [];
      }
    } else if (response is List) {
      list = response;
    }
    return list.map((c) => Complaint.fromJson(c as Map<String, dynamic>)).toList();
  }

  // ─── Municipality & Public ───

  Future<List<Complaint>> getMunicipalityComplaints({int limit = 20}) async {
    final response = await _apiClient.get(
      '/public/my-municipality-complaints?limit=$limit&status=VALIDATED,ASSIGNED,IN_PROGRESS,RESOLVED',
    );
    final list = response['complaints'] ?? [];
    return (list as List).map((c) => Complaint.fromJson(c)).toList();
  }

  Future<List<Complaint>> getRecentResolutions({int limit = 6}) async {
    final response = await _apiClient.get(
      '/public/my-municipality-complaints?limit=$limit&status=RESOLVED&sort=-updatedAt',
    );
    final list = response['complaints'] ?? [];
    return (list as List).map((c) => Complaint.fromJson(c)).toList();
  }

  Future<List<dynamic>> getTrendAlerts() async {
    try {
      final response = await _apiClient.get('/ai/trend/alerts');
      return response['data'] ?? [];
    } catch (_) {
      return [];
    }
  }

  // ─── Upload Media ───

  Future<List<dynamic>> uploadMedia(List<String> filePaths) async {
    // Simulate upload - in real app use multipart
    final response = await _apiClient.post('/upload', {
      'files': filePaths,
    });
    return response['data'] ?? [];
  }

  // ─── Notifications ───

  Future<List<dynamic>> getNotifications({int page = 1, int limit = 50}) async {
    final response = await _apiClient.get(
      '/notifications?page=$page&limit=$limit',
    );
    return response['notifications'] ?? response['data'] ?? [];
  }

  Future<int> getUnreadCount() async {
    final response = await _apiClient.get('/notifications/count');
    return response['unread'] ?? response['count'] ?? 0;
  }

  Future<void> markAsRead(String id) async {
    await _apiClient.put('/notifications/$id/read', {});
  }

  Future<void> markAllAsRead() async {
    await _apiClient.put('/notifications/read-all', {});
  }

  // Duplicate Detection (BL-25)
  Future<Map<String, dynamic>> checkDuplicate({
    required String title,
    required String description,
    required String category,
    required String municipality,
    double? latitude,
    double? longitude,
  }) async {
    final body = <String, dynamic>{
      'title': title,
      'description': description,
      'category': category,
      'municipality': municipality,
    };
    if (latitude != null) body['latitude'] = latitude;
    if (longitude != null) body['longitude'] = longitude;
    final response = await _apiClient.post('/ai/duplicate/check', body);
    return response;
  }

  Future<Map<String, dynamic>> confirmDuplicateDecision({
    required String newComplaintId,
    required String existingComplaintId,
    required String action,
  }) async {
    final response = await _apiClient.post('/ai/duplicate/confirm', {
      'newComplaintId': newComplaintId,
      'existingComplaintId': existingComplaintId,
      'action': action,
    });
    return response;
  }
}
